### ISSUE TITLE
Analyzing SG 147

### ISSUE THREAD

**Issue by samuelkim314** [thchr/MPBUtils.jl#18]
Is there anything different I need to do to analyze spacegroup 147, which has broken time reversal symmetry? My code looks something like:

```julia
has_tr = false
D = 3
symeigsd, lgd = read_symdata(calcname; sgnum=sgnum, D=D, dir="output/")
# --- fix singular photonic symmetry content at Γ, ω=0 ---
lgs = littlegroups(sgnum, Val(D)) 
fixup_gamma_symmetry!(symeigsd, lgs["Γ"])
lgirsd = pick_lgirreps(lgd)
brs = bandreps(sgnum, timereversal=has_tr)
# --- analyze connectivity and topology of symmetry data ---
summaries = analyze_symmetry_data(symeigsd, lgirsd, brs)
```

but I get the following error:

```julia
ERROR: Could not find irrep label A₂⁺A₃⁺
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] find_permutation(lgirsd::Dict{String, Vector{LGIrrep{3}}}, klabs::Vector{String}, irlabs::Vector{String})
   @ MPBUtils ~/.julia/dev/MPBUtils/src/read_utils.jl:297
 [3] find_permutation
   @ ~/.julia/dev/MPBUtils/src/read_utils.jl:282 [inlined]
 [4] _default_permutation
   @ ~/.julia/dev/MPBUtils/src/read_utils.jl:308 [inlined]
 [5] analyze_symmetry_data(symeigsd::Dict{String, Vector{Vector{ComplexF64}}}, lgirsd::Dict{String, Vector{LGIrrep{3}}}, brs::BandRepSet; multiplicities_kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ MPBUtils ~/.julia/dev/MPBUtils/src/read_utils.jl:89
 [6] analyze_symmetry_data(symeigsd::Dict{String, Vector{Vector{ComplexF64}}}, lgirsd::Dict{String, Vector{LGIrrep{3}}}, brs::BandRepSet)
   @ MPBUtils ~/.julia/dev/MPBUtils/src/read_utils.jl:83
 [7] checkhinge2(sgnum::Int64, calcname::String; topoindex::Vector{Int64})
   @ Main /mnt/c/Users/Sam/Documents/MIT/TopPhC/photonic-topo-sym_mpb-ctl/process_symeigs.jl:258
 [8] checkhinge2(sgnum::Int64, calcname::String)
   @ Main /mnt/c/Users/Sam/Documents/MIT/TopPhC/photonic-topo-sym_mpb-ctl/process_symeigs.jl:234
 [9] top-level scope
   @ REPL[8]:1
```

**Comment by thchr** [comment]
I think there's two issues:

1. `pick_lgirreps` should be given a `timereversal=has_tr` keyword argument; that's why you're seeing the "doubled-up" irrep A₂⁺A₃⁺ (which cannot exist without time-reversal). So `pick_lgirreps(lgd)` should become `pick_lgirreps(lgd; timereversal=has_tr)`.
We're missing some documentation for `pick_lgirreps`: the main function it has it to return `lgirreps` in the particular sorting of `lgd`'s operations (and implicitly, it also converts the operations in the returned `LGIrrep`s to a primitive setting if the kwarg `isprimitive` is `true` (the default)).
2. The line `lgs = littlegroups(sgnum, Val(D))` is subtly wrong, because it returns the little group operations in a conventional setting. For it to be passed to `fixup_gamma_symmetry!` it should be in the primitive setting (since that's what the `symeigsd` values were calculated for). In fact, because the `symeigsd` values were calculated for a particular operator sorting, which is the operator sorting of `lgd`, it'd be better to just pass `lgd["Γ"]` to `fixup_gamma_symmetry!`.

In summary, something like the following would be better:
```jl
timereversal = false
D = 3
symeigsd, lgd = read_symdata(calcname; sgnum=sgnum, D=D, dir="output/")
# --- fix singular photonic symmetry content at Γ, ω=0 ---
fixup_gamma_symmetry!(symeigsd, lgd["Γ"])
lgirsd = pick_lgirreps(lgd; timereversall)
brs = bandreps(sgnum, timereversal)
# --- analyze connectivity and topology of symmetry data ---
summaries = analyze_symmetry_data(symeigsd, lgirsd, brs)
```

If you see a way to improve the current README.md to make this implicitly obvious/avoid this pitfall, it'd be great with a PR.

**Comment by thchr** [comment]
Btw., if you would: can you confirm that the new methods (i.e. `analyze_symmetry_data`) and the old methods give the same answers for your SG147 calculations?
I haven't done any explicit comparisons myself.

... we really should add some tests to this package and set up CI...

**Comment by samuelkim314** [comment]
Thanks for the help! It runs now, but it gives different answer versus my old method.

So the old code is this:

```julia
"""Check symmetry eigenvalues to see if the bands allow for a hinge state"""
function checkhinge(sgnum, calcname; topoindex=[0, 6])
    # --- setup info ---
    has_tr = false

    # --- hilbert/ebr bases ---
    sb, brs = compatibility_basis(sgnum, 3; timereversal=has_tr)
    B = matrix(brs, true)
    F = smith(B)

    # --- load and process data ---
    bandirsd, lgirsd = extract_individual_multiplicities(
                            calcname,
                            timereversal=has_tr,
                            dir = "output/",
                            atol=2e-2,
                            D=3)
    length(lgirsd) ≠ length(sb.klabs) && error("missing k-point data")

    # extract the _potentially_ separable symmetry vectors `ns` and their band-ranges `bands`
    bands = nothing
    ns = nothing
    try
        bands, ns = extract_candidate_symmetryvectors(bandirsd, lgirsd, brs)
    catch e
        if e == ErrorException("found no isolable band candidates")
            @warn "found no isolable band candidates"
            return nothing
        else
            throw(e)
        end
    end

    # --- find which band combinations are in {BS} and then test their topology ---
    band′ = 0:0
    n′ = similar(first(ns))
    is_bs = true # at every new iteration, `is_bs` effectively means `was_prev_iter_bs`
    for (band, n) in zip(bands, ns)
        band′, n′, is_bs

        band′ = is_bs ? band : (minimum(band′):maximum(band))
        n′    = is_bs ? n : n′ + n

        # test if `n′` is in {BS} or not
        is_bs = if first(band′) == 1                        # singular bands
            is_transverse_bandstruct(n′, sb, lgirsd["Γ"], F)
        else                                                # regular bands
            isbandstruct(n′, F)
        end
        if !is_bs
            #println("   ", band′, " not a band structure")
            continue
        end

        # calculate topological index of `bands′`
        ν,_ = if first(band′) == 1
            indicators_singular(n′, sb, lgirsd["Γ"], B)
        else
            indicators(n′, F)
        end
        println((band′, ν))
        if ν == topoindex
            # We have a winner - possibly; although maybe we should also impose a check
            # on `band′``
            return band′
        end
    end

    return nothing
end
```
and it prints
`(1:6, [0, 6])`
and returns `1:6`

But the new code 
```julia
function checkhinge2(sgnum, calcname; topoindex=[0, 6])
    timereversal = false
    D = 3
    symeigsd, lgd = read_symdata(calcname; sgnum=sgnum, D=D, dir="output/")
    # --- fix singular photonic symmetry content at Γ, ω=0 ---
    fixup_gamma_symmetry!(symeigsd, lgd["Γ"])
    lgirsd = pick_lgirreps(lgd; timereversal)
    brs = bandreps(sgnum; timereversal=timereversal)
    # --- analyze connectivity and topology of symmetry data ---
    summaries = analyze_symmetry_data(symeigsd, lgirsd, brs)
    return summaries
end
```
and the `summaries` variable contains:
```julia
julia> summaries[1]
7-band BandSummary:
 bands:      1:7
 n:          2A₂⁺+2A₂⁻+2A₃⁺+A₃⁻, -Γ₁⁺+Γ₁⁻+2Γ₂⁺+2Γ₂⁻+Γ₃⁺+2Γ₃⁻, 3H₁+2H₂+2H₃, 3K₁+2K₂+2K₃, 3L₁⁺+4L₁⁻, 3M₁⁺+4M₁⁻
 topology:   nontrivial
 indicators: (1,8) ∈ Z₂×Z₁₂

julia> summaries[2]
1-band BandSummary:
 bands:      8:8
 n:          A₂⁺, Γ₂⁻, H₃, K₃, L₁⁻, M₁⁺
 topology:   nontrivial
 indicators: (1,4) ∈ Z₂×Z₁₂
```

**Comment by thchr** [comment]
Hmmm. Can you upload/paste the associated data files i need to reproduce this? Definitely want to make sure the old/new approaches agree.

**Comment by thchr** [comment]
Okay, this took me too long, but I figured the issue; the keyword `atol` was not being set explicitly in your "new" variations, and so it fell back to the default value (1e-2). In your "old" variation, you explicitly set `atol = 2e-2`.

Somehow, this difference is enough to make resolving the A irrep at the 6th and 7th band troublesome: 
1. For `atol = 1e-2`, the irreps at the 6th and 7th band *cannot* be individually resolved individually at the desired tolerance. Instead, it resolves to A₁⁻+A₃⁺ over bands 6 and 7 in combination.
2. For `atol = 2e-2`, the irreps now *can* be resolved individually, resolving to A₃⁺ for band 6 and to A₁⁻ for band 7.

The reason the `atol = 1e-2` case doesn't resolve is just that the accuracy of the symmetry eigenvalue calculation is a bit low. Specifically, the translation from symmetry eigenvalues to irreps gives the answer in floating point weights. It is this answer that is checked against `atol`. For the data you sent me, the answer for band 6 is `1.915e-5A₁⁺+0.014683763455326574A₁⁻+2.07e-5A₂⁺+0.0001876A₂⁻+0.984A₃⁺+0.000182A₃⁻`. Note that the coefficient of `A₃⁺` differs by more than `1e-2` from 1; hence the issue.

Long story short - the discrepancy is due to a different choice of `atol` and a relatively low resolution calculation of the symmetry eigenvalues. So there shouldn't be anything fundamentally wrong here.

**Comment by thchr** [comment]
For good measure, I get the following `summaries` at `atol = 2-e2` (seems unchanged if I change the value higher):
```jl
julia> summaries
3-element Vector{BandSummary}:
 6-band (nontrivial): [A₂⁺+2A₂⁻+2A₃⁺+A₃⁻, -Γ₁⁺+Γ₁⁻+Γ₂⁺+2Γ₂⁻+Γ₃⁺+2Γ₃⁻, 2H₁+2H₂+2H₃, 2K₁+2K₂+2K₃, 3L₁⁺+3L₁⁻, 3M₁⁺+3M₁⁻]
 1-band (nontrivial): [A₂⁺, Γ₂⁺, H₁, K₁, L₁⁻, M₁⁻]
 1-band (nontrivial): [A₂⁺, Γ₂⁻, H₃, K₃, L₁⁻, M₁⁺]

julia> getproperty.(summaries, :indicators)
3-element Vector{Vector{Int64}}:
 [0, 6]
 [1, 2]
 [1, 4]
```

**Comment by thchr** [comment]
I pushed https://github.com/thchr/MPBUtils.jl/commit/9fafd40753e9cfa1281e2e326f62a2923d48324f to increase the default `atol` to `2e-2` and also extend some of the associated doc-strings to hopefully better clarify this.

**Comment by samuelkim314** [comment]
Ohh interesting, thanks so much for looking into this! Do you think `atol=2e-2` should generally work moving forward, or do we need to consider tuning it further to avoid missing possible cases? Or going back to higher resolution calculations?The numerical discrepancy of ~`0.016` for `A₃⁺` seems rather large.
